src/Repository/Storable.php

```diff
@@ -0,0 +1,143 @@
+use Friendica\Database\Database;
+use Friendica\Database\DBA;
+use Friendica\Network\HTTPException;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Storable repositories are linked to a main table with an id primary key and will instanciate Models and Collections
+ * related to that table.
+ */
+abstract class Storable extends Repository
+{
+	const LIMIT = 30;
+
+	/** @var Database */
+	protected $dba;
+
+	/** @var string */
+	protected static $table_name;
+
+	/** @var Model */
+	protected static $model_class;
+
+	/** @var Collection */
+	protected static $collection_class;
+
+	public function __construct(Database $dba, LoggerInterface $logger)
+	{
+		parent::__construct($logger);
+
+		$this->dba = $dba;
+		$this->logger = $logger;
+	}
+
+	/**
+	 * Fetches a single model record. The condition array is expected to contain a unique index (primary or otherwise).
+	 *
+	 * Chainable.
+	 *
+	 * @param array $condition
+	 * @return Model
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectFirst(array $condition)
+	{
+		$data = $this->dba->selectFirst(static::$table_name, [], $condition);
+
+		if (!$data) {
+			throw new HTTPException\NotFoundException(static::class . ' record not found.');
+		}
+
+		return new static::$model_class($data);
+	}
+
+	/**
+	 * Populates a Collection according to the condition.
+	 *
+	 * Chainable.
+	 *
+	 * @param array $condition
+	 * @param array $params
+	 * @return Collection
+	 * @throws \Exception
+	 */
+	public function select(array $condition = [], array $params = [])
+	{
+		$result = $this->dba->select(static::$table_name, [], $condition, $params);
+
+		$models = [];
+
+		while ($record = $this->dba->fetch($result)) {
+			$models[] = new static::$model_class($this->dba, $this->logger, $record);
+		}
+
+		return new static::$collection_class($models);
+	}
+
+	/**
+	 * Populates the collection according to the condition. Retrieves a limited subset of models depending on the boundaries
+	 * and the limit. The total count of rows matching the condition is stored in the collection.
+	 *
+	 * Chainable.
+	 *
+	 * @param array $condition
+	 * @param array $params
+	 * @param int?  $max_id
+	 * @param int?  $since_id
+	 * @param int   $limit
+	 * @return $this
+	 * @throws \Exception
+	 */
+	public function selectByBoundaries(array $condition = [], array $params = [], int $max_id = null, int $since_id = null, int $limit = self::LIMIT)
+	{
+		$condition = DBA::collapseCondition($condition);
+
+		$boundCondition = $condition;
+
+		if (isset($max_id)) {
+			$boundCondition[0] .= " AND `id` < ?";
+			$boundCondition[] = $max_id;
+		}
+
+		if (isset($since_id)) {
+			$boundCondition[0] .= " AND `id` > ?";
+			$boundCondition[] = $since_id;
+		}
+
+		$params['limit'] = $limit;
+
+		$result = $this->dba->select(static::$table_name, [], $boundCondition, $params);
+
+		$models = [];
+		while ($record = $this->dba->fetch($result)) {
+			$models[] = new static::$model_class($this->dba, $this->logger, $record);
```

Comment:
This is why I use the `Prototype` pattern:
https://designpatternsphp.readthedocs.io/en/latest/Creational/Prototype/README.html

If we use this repository for - let's say - items, where you select 1.000 of items to get updated, cloning a pre-existing instance is faster than plain instancing. And that's why I would - in this case - make this one:
```suggestion
			$models[] = new static::$model_class::createFromPrototype($prototype, $record);
```
Where `$prototype` is the **first** instance, you created with this method and re-use it for the next ones :-)

Reply:
But then you have a static factory, I thought this was a no-no? You made me change my Mastodon entities to not have anymore static factories?

Reply:
In this case, it's a "wrapper" for:
```php
$newModel = clone $firstModel;
$newModel->attributes = $attributes;
```
it has no dependencies, that's what static factories make it problematic :-)

Reply:
Fair enough, but this is performance optimization that can be tacked on later visiby.